src/Endatix.Core/Specifications/SpecificationExtensions.cs

```diff
@@ -30,4 +34,110 @@ public static ISpecificationBuilder<TEntity> Paginate<TEntity>(this ISpecificati
         return query.Take(filter.PageSize);
     }
 
+    /// <summary>
+    /// Applies filtering to the query based on the provided filter parameters
+    /// </summary>
+    /// <typeparam name="TEntity">The type of entity being queried</typeparam>
+    /// <param name="query">The specification builder instance</param>
+    /// <param name="filters">The filter parameters to apply</param>
+    /// <returns>The filtered specification builder</returns>
+    public static ISpecificationBuilder<TEntity> Filter<TEntity>(this ISpecificationBuilder<TEntity> query, FilterParameters filters)
+    {
+        if (filters?.Criteria == null || !filters.Criteria.Any())
+        {
+            return query;
+        }
+
+        return filters.Criteria.Aggregate(query, (current, criterion) => current.Filter(criterion));
+    }
+
+    /// <summary>
+    /// Applies a single filter criterion to the query
+    /// </summary>
+    /// <typeparam name="TEntity">The type of entity being queried</typeparam>
+    /// <param name="query">The specification builder instance</param>
+    /// <param name="filter">The filter criterion to apply</param>
+    /// <returns>The filtered specification builder</returns>
+    public static ISpecificationBuilder<TEntity> Filter<TEntity>(this ISpecificationBuilder<TEntity> query, FilterCriterion filter)
+    {
+        if (filter.Values == null || !filter.Values.Any())
+        {
+            return query;
+        }
+
+        var entityType = typeof(TEntity);
+        var propertyInfo = entityType.GetProperty(
+            filter.Field, 
+            BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase);
+        
+        Guard.Against.Null(
+            propertyInfo,
+            nameof(filter.Field), 
+            $"Property '{filter.Field}' does not exist on type '{entityType.Name}'");
+

```

Comment:
I agree the reflection has its drawback. I'm not sure however, how much it is used around our code base and especially in the libraries we use. I think it's good to have some analysis and let the numbers speak before we put efforts for optimizations. For sure the reflection is slow but who knows it may turn out that compared to a single DB request is almost nothing and optimizing it could improve very little... just an example.
I don't really understand your example from the code you share.
Generally I would suggest we plan and start doing performance testing so we can analyze where are the bottlenecks, then really focus our efforts on what mostly slows our system.
As for caching - this really sounds good and will be an easy way to considerably lower the reflection calls. If you agree, let's start with this step. I can add a task for caching here.

Reply:
I have seen apps crashing because of the reflection overhead and hence me being proactive about it. Fully agree on the stress testing instead of blind premature optimizations - this is the way!

For the caching task, yes - please add it. It is an easy win here especially since we don't need cache invalidation (fields don't change unless the code change). For the task, please mark it with  [:zap: performance](https://github.com/endatix/endatix/labels/%3Azap%3A%20performance) and [tech-debt](https://github.com/endatix/endatix/labels/tech-debt).

Reply:
Here it is - https://github.com/endatix/endatix/issues/301